cmd/commands/controller.go

```diff
@@ -0,0 +1,326 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	appv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	argoevents "github.com/argoproj/argo-events"
+	"github.com/argoproj/argo-events/common"
+	"github.com/argoproj/argo-events/common/logging"
+	"github.com/argoproj/argo-events/controllers/eventbus"
+	"github.com/argoproj/argo-events/controllers/eventsource"
+	"github.com/argoproj/argo-events/controllers/sensor"
+	eventbusv1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1"
+	eventsourcev1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventsource/v1alpha1"
+	sensorv1alpha1 "github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1"
+	envpkg "github.com/argoproj/pkg/env"
+)
+
+const (
+	namespaceEnvVar           = "NAMESPACE"
+	natsStreamingEnvVar       = "NATS_STREAMING_IMAGE"
+	natsMetricsExporterEnvVar = "NATS_METRICS_EXPORTER_IMAGE"
+	eventSourceImageEnvVar    = "EVENTSOURCE_IMAGE"
+	sensorImageEnvVar         = "SENSOR_IMAGE"
+
+	componentValEventBus    = "eventbus"
+	componentValEventSource = "eventsource"
+	componentValSensor      = "sensor"
+)
+
+func NewControllerCommand() *cobra.Command {
+	var (
+		component        string
+		namespaced       bool
+		managedNamespace string
+	)
+
+	command := &cobra.Command{
+		Use:   common.ControllerCommand,
+		Short: "Start a controller",
+		Run: func(cmd *cobra.Command, args []string) {
+			switch component {
+			case componentValEventBus:
+				startEventBusController(namespaced, managedNamespace)
+			case componentValEventSource:
+				startEventSourceController(namespaced, managedNamespace)
+			case componentValSensor:
+				startSensorController(namespaced, managedNamespace)
+			case "":
+				cmd.HelpFunc()(cmd, args)
+			default:
+				fmt.Printf("Error: Invalid component: %s\n\n", component)
+				cmd.HelpFunc()(cmd, args)
+			}
+		},
+	}
+	defaultNamespace := envpkg.LookupEnvStringOr(namespaceEnvVar, "argo-events")
+	command.Flags().StringVar(&component, "component", "", `The controller component, possible values: "`+componentValEventBus+`", "`+componentValEventSource+`" or "`+componentValSensor+`".`)
+	command.Flags().BoolVar(&namespaced, "namespaced", false, "Whether to run in namespaced scope, defaults to false.")
+	command.Flags().StringVar(&managedNamespace, "managed-namespaces", defaultNamespace, "The namespace that the controller watches when \"--namespaced\" is \"true\".")
+	return command
+}
+
+func startEventBusController(namespaced bool, managedNamespace string) {
```

Comment:
Original eventbus-controller main(), almost no code change.

Reply:
I think you should put each sub-command it its own file

Reply:
done.